File: 797.所有可能的路径.py
#
# @lc app=leetcode.cn id=797 lang=python3
#
# [797] 所有可能的路径
#

# @lc code=start
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        res,path=[],[0]
        def dfs(index):
            if index==len(graph)-1:
                res.append(path[:])
            for x in graph[index]:
                path.append(x)
                dfs(x)
                path.pop()
        dfs(0)
        return res    
# @lc code=end

# Solution1：
# class Solution:
#     def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
#         res=[]
#         def dfs(index,path):
#             if index==len(graph)-1:
#                 res.append(path)
#             for x in graph[index]:
#                 p=path.copy()
#                 p.append(x)
#                 dfs(x,p)
#         dfs(0,[0])
#         return res   

# class Solution:
#     def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
#         res,path=[],[0]
#         def dfs(index):
#             if index==len(graph)-1:
#                 res.append(path[:])
#             for x in graph[index]:
#                 path.append(x)
#                 dfs(x)
#                 path.pop()
#         dfs(0)
#         return res    

File: review/797.所有可能的路径.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=797 lang=java
 *
 * [797] 所有可能的路径
 */

// @lc code=start
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> res=new LinkedList<>();
        List<Integer> list=new ArrayList<>(){{
            add(0);
        }};
        dfs(res,graph,list,0);
        return  res;
    }
    public void dfs(List<List<Integer>> res,int[][] graph,List<Integer> list,int index){
        if(index==graph.length-1){
            res.add(new ArrayList<>(list));
            return;
        }
        for(int i:graph[index]){
            list.add(i);
            dfs(res, graph, list, i);
            list.remove(list.size()-1);
        }
    }
}
// @lc code=end

// Solution1:
// class Solution {
//     public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
//         List<List<Integer>> res=new LinkedList<>();
//         List<Integer> list=new ArrayList<>(){{
//             add(0);
//         }};
//         dfs(res,graph,list,0);
//         return  res;
//     }
//     public void dfs(List<List<Integer>> res,int[][] graph,List<Integer> list,int index){
//         if(index==graph.length-1){
//             res.add(new ArrayList<>(list));
//             return;
//         }
//         for(int i:graph[index]){
//             list.add(i);
//             dfs(res, graph, list, i);
//             list.remove(list.size()-1);
//         }
//     }
// }